### PR TITLE
Lexer can now scan names, hex sequences, brackets and minus signs.

### PR DESCRIPTION
**Pull request by angusholder**
#24 
Not complete yet, I want to add some tests, just putting this here so you can see what it looks like so far.

### CODE REVIEW

File: lib/reader/src/lexer.rs [hunk 1]
```diff
@@ -326,6 +338,39 @@ impl<'a> Lexer<'a> {
...
     }
 
+    fn scan_name(&mut self) -> Result<LocatedToken<'a>, LocatedError> {
+        let loc = self.loc();
+        let begin = self.pos;
+
+        assert!(self.lookahead == Some('%'));
+
+        while let Some(c) = self.next_ch() {
+            if !c.is_alphanumeric() && c != '_' {
```
Comment (stoklund) [inline lib/reader/src/lexer.rs]: I think in this specific case, we should restrict to ASCII characters only (i.e., reject other Unicode chars). The `FuncName` printer will do the same.

It's not that important if Cretonne accepts Unicode names or not. The `.cton` files are just for writing tests.

File: lib/reader/src/lexer.rs [hunk 2]
```diff
@@ -326,6 +338,39 @@ impl<'a> Lexer<'a> {
...
+
+    fn scan_hex_sequence(&mut self) -> Result<LocatedToken<'a>, LocatedError> {
+        let loc = self.loc();
+        let begin = self.pos;
+
+        assert!(self.lookahead == Some('#'));
+
+        while let Some(c) = self.next_ch() {
+            match c {
+                'a'...'f' | 'A'...'F' | '0'...'9' => {},
```
Comment (stoklund) [inline lib/reader/src/lexer.rs]: See `char::is_digit()`.

**Comment by stoklund** [comment]
Looking good so far.

File: lib/reader/src/lexer.rs (hunk 1 above)
Comment (angusholder) [inline lib/reader/src/lexer.rs]: I can't seem to find any stdlib functions to check for ascii-only alphabetic characters, you don't know if there are any do you?

Comment (stoklund) [inline lib/reader/src/lexer.rs]: There's an example in `ir/funcname.rs`.

**Comment by angusholder** [comment]
I'm fairly happy with the lexer as-is now, do you want me to make separate pull request for the parser work then?

**Comment by stoklund** [comment]
> I'm fairly happy with the lexer as-is now, do you want me to make separate pull request for the parser work then?
> 
Thanks, I'll review it in the morning.

Go ahead and start a new PR for the parser.

**Pull request merged**